toolkit/tools/internal/randomization/randomization_test.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package randomization
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"microsoft.com/pkggen/internal/logger"
+)
+
+const (
+	normalString    = "abcde12345"
+	multiByteString = "世界您好IsHelloWorld"
+	emojiString     = "1🤷‍♂️2@3( •_•)>⌐■~■ab~52(⌐■_■)67👩‍💻"
+)
+
+func TestMain(m *testing.M) {
+	testResult := 0
+
+	logger.InitStderrLog()
+	testResult = m.Run()
+
+	os.Exit(testResult)
+}
+
+func TestRandomNoNormalStringError(t *testing.T) {
+	var (
+		inputString = normalString
+		length      = 100
+	)
+	randomString, err := RandomString(length, inputString)
+	assert.NoError(t, err)
+	// We need to count runes, not
+	assert.Equal(t, length, len([]rune(randomString)))
+
+	fmt.Println(randomString)
+	fmt.Printf("len: %d\n", len(randomString))
+
+	// Heuristic to check we are getting a "roughly" random distribution:
+	// Generate random strings, check if each character is used at least an (average / 2) number of times at some point during the trials
+	// ie 100/10 = 10, 10/2 = 5, make sure every character shows up at least 5 times during the repeated trials
+	goodReturn := make(map[rune]bool)
+	for trial := 0; trial < 10; trial++ {
+		//trialSuccessful := true
+		randomString, err := RandomString(length, inputString)
+		assert.NoError(t, err)
+		for _, c := range inputString {
+			count := strings.Count(randomString, string(c))
+			if count > (length/len(inputString))/2 {
+				goodReturn[c] = true
+			}
+		}
+	}
+
+	for _, c := range inputString {
+		assert.Truef(t, goodReturn[c], "Char %c did not show up expected number of times in random string", c)
+	}
```

Comment:
I think we shouldn't do tests, which are based on probability. Every now and then we might get a failure for now good reason. If anything, I'd suggest we check, that the returned string if of the right length (rune-wise) and that the runes are all from the input set.

Reply:
I did a bit of looking, apparently something like this is considered a good way to validate. You just crank the number of test passes high enough so it "never" fails.

I'll add a test to make sure I'm not getting any incorrect runes in the output though.

---
In reply to: [549905569](https://github.com/microsoft/CBL-Mariner/pull/479#discussion_r549905569) [](ancestors = 549905569)

Reply:
I now validate that our test cases are guaranteed (prob of failure < 1e-100) to work.

---
In reply to: [550333756](https://github.com/microsoft/CBL-Mariner/pull/479#discussion_r550333756) [](ancestors = 550333756,549905569)